packages/interactivity/src/proxies/state.ts

```diff
@@ -0,0 +1,238 @@
+			} else {
+				prop.setValue(
+					shouldProxy( value ) ? proxifyState( ns, value ) : value
+				);
+			}
+		}
+	}
+	return props.get( key )!;
+};
+
+/**
+ * Relates each proxied object (i.e., the original object) with a signal that
+ * tracks changes in the number of properties.
+ */
+const objToIterable = new WeakMap< object, Signal< number > >();
+
+/**
+ * When this flag is `true`, it avoids any signal subscription, overriding state
+ * props' "reactive" behavior.
+ */
+let peeking = false;
+
+/**
+ * Handlers for reactive objects and arrays in the state.
+ */
+const stateHandlers: ProxyHandler< object > = {
+	get( target: object, key: string | symbol, receiver: object ): any {
+		/*
+		 * The property should not be reactive for the following cases:
+		 * 1. While using the `peek` function to read the property.
+		 * 2. The property exists but comes from the Object or Array prototypes.
+		 * 3. The property key is a known symbol.
+		 */
+		if (
+			peeking ||
+			( ! target.hasOwnProperty( key ) && key in target ) ||
+			( typeof key === 'symbol' && wellKnownSymbols.has( key ) )
+		) {
+			return Reflect.get( target, key, receiver );
+		}
+
+		// At this point, the property should be reactive.
+		const desc = Object.getOwnPropertyDescriptor( target, key );
+		const prop = getPropSignal( receiver, key, desc );
+
+		if ( peeking ) {
+			return prop.getComputed().peek();
+		}
+
+		const result = prop.getComputed().value;
+
+		/*
+		 * Check if the property is a synchronous function. If it is, set the
+		 * default namespace. Synchronous functions always run in the proper scope,
+		 * which is set by the Directives component.
+		 */
+		if ( typeof result === 'function' ) {
+			const ns = getProxyNs( receiver );
+			return ( ...args: unknown[] ) => {
+				setNamespace( ns );
+				try {
+					return result.call( receiver, ...args );
+				} finally {
+					resetNamespace();
+				}
+			};
+		}
+
+		return result;
+	},
+
+	set(
+		target: object,
+		key: string,
+		value: unknown,
+		receiver: object
+	): boolean {
+		setNamespace( getProxyNs( receiver ) );
+		try {
+			return Reflect.set( target, key, value, receiver );
+		} finally {
+			resetNamespace();
+		}
+	},
+
+	defineProperty(
+		target: object,
+		key: string,
+		desc: PropertyDescriptor
+	): boolean {
+		const isNew = ! ( key in target );
+		const result = Reflect.defineProperty( target, key, desc );
+
+		if ( result ) {
+			const receiver = getProxy( target );
+			const prop = getPropSignal( receiver, key );
+			const { get, value } = desc;
+			if ( get ) {
+				prop.setGetter( desc.get! );
+			} else {
+				const ns = getProxyNs( receiver );
+				prop.setValue(
+					shouldProxy( value ) ? proxifyState( ns, value ) : value
+				);
+			}
+
+			if ( isNew && objToIterable.has( target ) ) {
+				objToIterable.get( target )!.value++;
+			}
+
+			if ( Array.isArray( target ) ) {
+				const length = getPropSignal( receiver, 'length' );
+				length.setValue( target.length );
+			}
```

Comment:
We don't need to update to `length` if no one is subscribed.

Something like this?

```suggestion
			if ( Array.isArray( target ) && getPropSignal( receiver, 'length' ) ) {
				getPropSignal( receiver, 'length' ).setValue( target.length );
			}
```

Reply:
Well, `getPropSignal` automatically creates a `PropSignal` instance, but I haven't implemented a way to know whether that `PropSignal` has been accessed. 😅 

I can think of two options:

1. Implement a `hasPropSignal` function that tells whether a `PropSignal` exists for the given property. It should be as simple as this:
   ```js
   const hasPropSignal = ( proxy, key ) => proxyToProps.get( proxy )?.has( key );
   ```
2. Implement a `hasBeenAccessed` flag inside `PropSignal` instances that become true once `getComputed()` is called.

---

PS: Note that a `PropSignal`'s internal signals are not instantiated until either `getComputed()`, `setValue()`, or `setGetter()` are called.

Reply:
I used the first option (doing the check directly in the conditional). Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/3eea77152f4cb4163c5b5502f0ae525eea697c70.

Reply:
Perfect, thank you.